pinot-core/src/test/java/com/linkedin/pinot/queries/FastHllQueriesTest.java

```diff
@@ -0,0 +1,257 @@
+  }
+
+  @Test
+  public void testFastHllWithLog2m8() throws Exception {
+    buildSegment(8);
+
+    // Test inner segment queries
+    // Test query without filter
+    AggregationOperator aggregationOperator = getOperatorForQuery(BASE_QUERY);
+    IntermediateResultsBlock resultsBlock = (IntermediateResultsBlock) aggregationOperator.nextBlock();
+    ExecutionStatistics executionStatistics = aggregationOperator.getExecutionStatistics();
+    Assert.assertEquals(executionStatistics.getNumDocsScanned(), 1L);
+    Assert.assertEquals(executionStatistics.getNumEntriesScannedInFilter(), 0L);
+    Assert.assertEquals(executionStatistics.getNumEntriesScannedPostFilter(), 2L);
+    Assert.assertEquals(executionStatistics.getNumTotalRawDocs(), 30000L);
+    List<Object> aggregationResult = resultsBlock.getAggregationResult();
+    Assert.assertEquals(((HyperLogLog) aggregationResult.get(0)).cardinality(), 24L);
+    Assert.assertEquals(((HyperLogLog) aggregationResult.get(1)).cardinality(), 1366L);
+    // Test query with filter
+    aggregationOperator = getOperatorForQueryWithFilter(BASE_QUERY);
+    resultsBlock = (IntermediateResultsBlock) aggregationOperator.nextBlock();
+    executionStatistics = aggregationOperator.getExecutionStatistics();
+    Assert.assertEquals(executionStatistics.getNumDocsScanned(), 6129L);
+    Assert.assertEquals(executionStatistics.getNumEntriesScannedInFilter(), 85248L);
+    Assert.assertEquals(executionStatistics.getNumEntriesScannedPostFilter(), 12258L);
+    Assert.assertEquals(executionStatistics.getNumTotalRawDocs(), 30000L);
+    aggregationResult = resultsBlock.getAggregationResult();
+    Assert.assertEquals(((HyperLogLog) aggregationResult.get(0)).cardinality(), 17L);
+    Assert.assertEquals(((HyperLogLog) aggregationResult.get(1)).cardinality(), 1011L);
+    // Test query with group-by
+    AggregationGroupByOperator aggregationGroupByOperator = getOperatorForQuery(BASE_QUERY + GROUP_BY);
+    resultsBlock = (IntermediateResultsBlock) aggregationGroupByOperator.nextBlock();
+    executionStatistics = aggregationGroupByOperator.getExecutionStatistics();
+    Assert.assertEquals(executionStatistics.getNumDocsScanned(), 4613L);
+    Assert.assertEquals(executionStatistics.getNumEntriesScannedInFilter(), 0L);
+    Assert.assertEquals(executionStatistics.getNumEntriesScannedPostFilter(), 13839L);
+    Assert.assertEquals(executionStatistics.getNumTotalRawDocs(), 30000L);
+    AggregationGroupByResult aggregationGroupByResult = resultsBlock.getAggregationGroupByResult();
+    GroupKeyGenerator.GroupKey firstGroupKey = aggregationGroupByResult.getGroupKeyIterator().next();
+    Assert.assertEquals(firstGroupKey.getStringKey(), "");
+    Assert.assertEquals(((HyperLogLog) aggregationGroupByResult.getResultForKey(firstGroupKey, 0)).cardinality(), 23L);
+    Assert.assertEquals(((HyperLogLog) aggregationGroupByResult.getResultForKey(firstGroupKey, 1)).cardinality(), 596L);
+
+    // Test inter segment queries
```

Comment:
This comment is misleading. We are testing queries on a single segment only. See comment below for PregeneratedHllTest about preserving test functionality for combining results across segments.